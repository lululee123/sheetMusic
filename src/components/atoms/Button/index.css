/* stylelint-disable no-descending-specificity */
.button {
	display: flex;
	justify-content: center;
	align-items: center;
	flex-shrink: 0;
	cursor: pointer;
	background-color: transparent;
	padding: 0;
	border: none;
	transition: opacity 0.3s ease;

	& > .main {
		display: inline-flex;
		align-items: center;
		justify-content: center;
		flex: 1;
		flex-shrink: 0;
		padding: 0 22.5px;
		border-radius: 2px;
		border-top: 1px solid var(--blue);
		border-bottom: 1px solid var(--blue);
		background-color: var(--blue);
		color: white;
		transition: all 0.3s ease;

		& > span {
			font-size: 16px;
			font-weight: 500;
			font-stretch: normal;
			font-style: normal;
			line-height: 1.5;
			letter-spacing: normal;
			text-align: center;
		}
	}

	& > .feather-icon {
		display: inline-flex;
		flex-shrink: 0;
		height: 100%;

		&.large > svg {
			height: 48px;
		}

		&.normal > svg {
			height: 40px;
		}

		&.small > svg {
			height: 32px;
		}

		& > svg {
			width: auto;
			margin-left: -1px;

			& > g {
				fill: var(--blue);
				transition: fill 0.3s ease;

				& path {
					stroke: var(--blue);
					transition: stroke 0.3s ease;
				}
			}
		}

		&.outline {
			& > svg g {
				fill: transparent;

				& path {
					stroke: var(--blue);
				}
			}
		}
	}

	&.large {
		& > .main {
			height: 48px;
		}
	}

	&.normal {
		& > .main {
			height: 40px;
		}
	}

	&.small {
		& > .main {
			height: 32px;

			& > span {
				font-size: 14px;
			}
		}
	}

	&.feather {
		& > .main {
			padding: 10.5px 0;
			border-radius: 0;
		}
	}

	&.outline {
		& > .main {
			border-radius: 0;
			background-color: transparent;
			color: var(--blue);
		}
	}

	&.link {
		& > .main {
			border-radius: 0;
			border: none;
			padding: 0;
			height: auto;
			background-color: transparent;
			color: var(--blue);
		}
	}

	&.white {
		& > .main {
			border-top: 1px solid white;
			border-bottom: 1px solid white;
			background-color: white;
			color: var(--blue);
		}

		& > .feather-icon {
			& > svg g {
				fill: white;

				& path {
					stroke: white;
				}
			}

			&.outline {
				& > svg g {
					& path {
						stroke: white;
					}
				}
			}
		}

		&.link {
			& > .main {
				border: none;
				background-color: transparent;
				color: white;
			}
		}
	}

	&.disabled {
		cursor: not-allowed;

		& > .main {
			border-top: 1px solid var(--lightBlue);
			border-bottom: 1px solid var(--lightBlue);
			background-color: var(--lightBlue);
		}

		&.link {
			& > .main {
				border: none;
				background-color: transparent;
				color: var(--lightBlue);
			}
		}

		& > .feather-icon {
			& > svg g {
				fill: var(--lightBlue);

				& path {
					stroke: var(--lightBlue);
				}
			}

			&.outline {
				& > svg g {
					& path {
						stroke: var(--lightBlue);
					}
				}
			}
		}
	}

	&:hover {
		& > .main {
			border-top: 1px solid var(--darkBlue);
			border-bottom: 1px solid var(--darkBlue);
			background-color: var(--darkBlue);
		}

		& > .feather-icon {
			& > svg {
				& > g {
					fill: var(--darkBlue);

					& path {
						stroke: var(--darkBlue);
					}
				}
			}

			&.outline {
				& > svg g {
					fill: transparent;

					& path {
						stroke: var(--darkBlue);
					}
				}
			}
		}

		&.outline {
			& > .main {
				color: var(--darkBlue);
				background-color: transparent;
			}
		}

		&.link {
			& > .main {
				color: var(--darkBlue);
				background-color: transparent;
				border: none;
			}
		}

		&.white {
			& > .main {
				border-top: 1px solid white;
				border-bottom: 1px solid white;
				background-color: white;
				color: var(--blue);
			}

			& > .feather-icon {
				& > svg g {
					fill: white;

					& path {
						stroke: white;
					}
				}

				&.outline {
					& > svg g {
						& path {
							stroke: white;
						}
					}
				}
			}

			&.link {
				& > .main {
					border: none;
					background-color: transparent;
					color: white;
				}
			}
		}

		&.disabled {
			& > .main {
				border-top: 1px solid var(--lightBlue);
				border-bottom: 1px solid var(--lightBlue);
				background-color: var(--lightBlue);
			}

			&.link {
				& > .main {
					color: var(--lightBlue);
				}
			}

			& > .feather-icon {
				& > svg g {
					fill: var(--lightBlue);

					& path {
						stroke: var(--lightBlue);
					}
				}

				&.outline {
					& > svg g {
						& path {
							stroke: var(--lightBlue);
						}
					}
				}
			}
		}
	}
}

/* stylelint-enable no-descending-specificity */
