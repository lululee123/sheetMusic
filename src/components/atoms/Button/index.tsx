import React, { MouseEvent } from 'react';
import classnames from 'classnames';

import { isExist } from 'util/helper';

import FeatherLarge from 'images/icon/feather-right-large.inline.svg';

import styles from './index.css';

enum SideType {
	left = 'left',
	right = 'right',
}

interface FeatherIconProperty {
	className?: string;
	size?: 'large' | 'normal' | 'small';
	outline?: boolean;
	disabled?: boolean;
	side?: SideType;
}

const FeatherIcon: React.FC<FeatherIconProperty> = ({
	className,
	size = 'normal',
	outline = false,
	disabled = false,
	side = SideType.right,
}) => (
	<div
		className={classnames(
			styles.featherIcon,
			className,
			{
				[styles.outline]: outline,
				[styles.disabled]: disabled,
			},
			isExist(size) && styles[size],
		)}
		style={{
			transform: side === SideType.left ? 'rotate(180deg)' : 'rotate(0deg)',
		}}
	>
		<FeatherLarge />
	</div>
);

export interface ButtonProperty {
	className?: string;
	textClassName?: string;
	type?: 'normal' | 'outline' | 'link';
	disabled?: boolean;
	feather?: boolean;
	size?: 'normal' | 'large' | 'small';
	onClick?: (e: MouseEvent) => void;
	color?: string;
}

const Button: React.FC<ButtonProperty> = ({
	className,
	textClassName,
	type = 'normal',
	disabled = false,
	feather = false,
	size = 'normal',
	color = 'blue',
	children,
	onClick = () => {},
	...props
}) => (
	<button
		className={classnames(
			styles.button,
			className,
			{
				[styles.outline]: type === 'outline',
				[styles.link]: type === 'link',
				[styles.disabled]: disabled && feather,
				[styles.feather]: feather,
				[styles.large]: size === 'large',
				[styles.normal]: size === 'normal',
				[styles.small]: size === 'small',
			},
			styles[color],
		)}
		type="button"
		onClick={e => {
			if (!disabled) {
				onClick(e);
			}
		}}
		// eslint-disable-next-line react/jsx-props-no-spreading
		{...props}
	>
		{type !== 'link' && feather && (
			<FeatherIcon
				size={size}
				outline={type === 'outline'}
				disabled={disabled}
				side={SideType.left}
			/>
		)}
		<div className={classnames(styles.main, textClassName)}>
			<span>{children}</span>
		</div>
		{type !== 'link' && feather && (
			<FeatherIcon
				size={size}
				outline={type === 'outline'}
				disabled={disabled}
				side={SideType.right}
			/>
		)}
	</button>
);

export default Button;
