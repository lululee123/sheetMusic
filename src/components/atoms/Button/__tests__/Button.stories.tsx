import React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean, select } from '@storybook/addon-knobs';

import Button from 'components/atoms/Button';

const stories = storiesOf('atoms/Button', module);

stories.add('__interactive', () => (
	<Button
		disabled={boolean('disabled', false)}
		feather={boolean('feather', true)}
		type={select('type', ['normal', 'outline', 'link'], 'normal')}
		size={select('size', ['large', 'normal', 'small'], 'normal')}
		color={select('color', ['blue', 'white'], 'blue')}
	>
		下一步
	</Button>
));

stories.add('with samll feather', () => (
	<Button disabled={boolean('disabled', false)} feather={boolean('feather', true)}>
		下一步
	</Button>
));
