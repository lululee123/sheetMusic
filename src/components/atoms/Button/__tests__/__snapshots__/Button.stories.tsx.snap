// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`__interactive 1`] = `
<DocumentFragment>
  <button
    class="button feather normal blue"
    type="button"
  >
    <div
      class="featherIcon normal"
      style="transform: rotate(180deg);"
    >
      <test-file-stub />
    </div>
    <div
      class="main"
    >
      <span>
        下一步
      </span>
    </div>
    <div
      class="featherIcon normal"
      style="transform: rotate(0deg);"
    >
      <test-file-stub />
    </div>
  </button>
</DocumentFragment>
`;

exports[`with samll feather 1`] = `
<DocumentFragment>
  <button
    class="button feather normal blue"
    type="button"
  >
    <div
      class="featherIcon normal"
      style="transform: rotate(180deg);"
    >
      <test-file-stub />
    </div>
    <div
      class="main"
    >
      <span>
        下一步
      </span>
    </div>
    <div
      class="featherIcon normal"
      style="transform: rotate(0deg);"
    >
      <test-file-stub />
    </div>
  </button>
</DocumentFragment>
`;
